### PR TITLE
💫 Update matcher engine for regex and extensions

### LINKED ISSUE #1971: 💫 Better, faster and more customisable matcher
Related issues: #1567, #1711, #1819, #1939, #1945, #1951, #2042

We're currently in the process of rewriting the match loop, fixing long-standing issues and making it easier to extend the `Matcher` and `PhraseMatcher`. The community contributions by @GregDubbin and @savkov have already made a big difference – we can't wait to get it all ready and shipped.

This issue discusses some of the planned new features and additions to the match patterns API, including matching by **custom extension attributes** (`Token._.`), **regular expressions**, **set membership** and **rich comparison for numeric values**.

### Custom extension attributes
spaCy v2.0 introduced [custom extension attributes](https://spacy.io/usage/processing-pipelines#custom-components-attributes) on the `Doc`, `Span` and `Token`. Custom attributes make it easier to attach arbitrary data to the built-in objects, and let users take advantage of spaCy's data structures and the `Doc` object as the "single source of truth". However, not being able to match on custom attributes was quite limiting (see #1499, #1825).

The new patterns spec will allow an `_` space on token patterns, which can map to a dictionary keyed by the attribute names:

```python
Token.set_extension('is_fruit', getter=lambda token: token.text in ('apple', 'banana'))

pattern = [{'LEMMA': 'have'}, {'_': {'is_fruit': True}}]
matcher.add('HAVING_FRUIT', None, pattern)
```

Both regular attribute extensions (with a default value) and property extensions (with a getter) will be supported and can be combined for more exact matches.

```python
pattern = [{'_': {'is_fruit': True, 'fruit_color': 'red', 'fruit_rating': 5}}]
```

### Rich comparison for numeric values
Token patterns already allow specifying a `LENGTH` (the token's character length). However, matching tokens of between five and ten characters previously required adding 6 copies of the exact same pattern, introducing unnecessary overhead. Numeric attributes can now also specify a dictionary with the predicate (e.g. `'>'` or `'<='`) mapped to the value. For example:

```python
pattern = [{'ENT_TYPE': 'ORG', 'LENGTH': 5}]          # exact length
pattern = [{'ENT_TYPE': 'ORG', 'LENGTH': {'>=': 5}}]  # length with predicate
```

The second pattern above will match a token with the entity type `ORG` that's 5 or more characters long. Combined with custom attributes, this allows very powerful queries combining both linguistic features and numeric data:

```python
# match a token based on custom numeric attributes
pattern = [{'_': {'fruit_rating': {'>': 7}, 'fruit_weight': {'>=': 100, '<': 300}}]

# match a verb with ._.sentiment_score >= 5 and one token on each side
pattern = [{}, {'POS': 'VERB', '_': {'sentiment_score': {'>=': 0.5}}}, {}]
```

Defining predicates and values as a dictionary instead of a single string like `'>=5'` allows us to avoid string parsing, and lets spaCy handle custom attributes without requiring the user to specify their types upfront. (While we know the type of the built-in `LENGTH` attribute, spaCy has no way of knowing whether the value `'<3'` of a custom attribute should be interpreted as "less than 3", or the heart emoticon.)

### Set membership
This is another feature that has been requested before and will now be much easier to implement. Similar to the predicate mapping for numeric values, token attributes can now also be defined as dictionaries. The keys `IN` or `NOT_IN` can be used to indicate set membership and non-membership.

```python
pattern = [{'LEMMA': {'IN': ['like', 'love']}}, 
           {'LOWER': {'IN': ['apples', 'bananas']}}]
```

The above pattern will match a token with the lemma "like" or "love", followed by a token whose lowercase form is either "apples" or "bananas". For example, "loving apples" or "likes bananas". Lists can be used for all non-boolean values, including custom `_` attributes:

```python
# verb or conjunction followed by custom is_fruit token
pattern = [{'POS': {'IN': ['VERB', 'CONJ', 'CCONJ']}}, 
           {'_': {'is_fruit': True, 'fruit_color': {'NOT_IN': ['red', 'yellow']}}}]

# set membership of numeric custom attributes
pattern = [{'_': {'is_customer': True, 'customer_year': {'IN': [2018, 2017, 2016]}}}]

# combination of predicates and and non-membership
pattern = [{'_': {'custom_count': {'<=': 100, 'NOT_IN': [12, 66, 79]}}}]
```

### Regular expressions
Using regular expressions within token patterns is [already possible via custom binary flags](https://spacy.io/usage/linguistic-features#regex) (see #1567). However, this has some inconvenient limitations – including the patterns not being JSON-serializable. If the solution is to add binary flags, spaCy might as well take care of that. The following example is based on the work by @savkov (see #1833):

```python
pattern = [{'ORTH': {'REGEX': '^([Uu](\\.?|nited) ?[Ss](\\.?|tates)'}},
           {'LOWER': 'president'}]
```

`'REGEX'` as an operator (instead of a top-level property that only matches on the token's text) allows defining rules for any string value, including custom attributes:

```python
# match tokens with fine-grained POS tags starting with 'V'
pattern = [{'TAG': {'REGEX': '^V'}}]

# match custom attribute values with regular expressions
pattern = [{'_': {'country': {'REGEX': '^([Uu](\\.?|nited) ?[Ss](\\.?|tates)'}}}]
```

### New operators

TL;DR: The new patterns spec will allow two ways of defining properties – attribute values for exact matches and dictionaries using operators for more fine-grained matches.

```python
{
    PROPERTY: value,                  # exact match
    PROPERTY: {OPERATOR: value, ...}  # match with operators
}
```

The following operators can be used within dictionaries describing attribute values:

| Operator | Value type | Description | Example |
| --- | --- | --- | --- |
| `==`, `>=`, `<=`, `>`, `<` | int, float | Attribute value is equal, greater or equal, smaller or equal, greater or smaller. | `'LENGTH': {'>': 10}` | 
| `IN` | any | Attribute value is member of a list. | `'LEMMA': {'IN': ['like', 'love']}` |
| `NOT_IN` | any | Attribute value is *not* member of a list. | `'POS': {'NOT_IN': ['NOUN', 'PROPN']}` |
| `REGEX` | unicode | Attribute value matches regular expression. | `'TAG': {'REGEX': '^V'}` |

## API improvements and bug fixes
See @honnibal's comments in #1945 and the [`feature/better-faster-matcher`](https://github.com/explosion/spaCy/tree/feature/better-faster-matcher) branch for more details and implementation examples.

### Other fixes
- [x] #1711: Remove hard-coded length limit of 10 on the `PhraseMatcher`.
- [x] #1939: Fix pickling of `PhraseMatcher`.
- [x] #1951: `Matcher.pipe` should yield matches instead of `Doc` objects.
- [ ] #2042: Support deleting rules in `PhraseMatcher`.
- [x] Accept `"TEXT"` as an alternative to `"ORTH"` (for consistency).

### PR DESCRIPTION
**Pull request by honnibal**
Add support for matching over arbitrary Python predicate functions, and
arbitrary Python attribute getters. This will allow matching over regex
patterns, and allow supporting extension attributes.

The results of the Python predicate functions are cached, so that we don't
call the same predicate function twice for the same token. The extension
attributes are fetched into an array for each token in the doc. This
should minimise the performance impact of the new features.

This patch tips the `matcher.pyx` file over the 1000 line mark :(. We should make
it a subpackage. We also need better pattern validation and errors.

Closes #1971

## Checklist

- [x] I have submitted the spaCy Contributor Agreement.
- [x] I ran the tests, and all new and existing tests passed.
- [ ] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

**Comment by honnibal** [comment]
@ines Everything should be working here, but the code's pretty rushed. In particular, I'm not doing anything to verify that the patterns are reasonable, so we'll get things dying in random places. All of the following are invalid for various reasons:

* `{"LENGTH": {"REGEX": "[4-9]"}}` --  Regex requires string valued attribute, length is int valued.
* `{"ORTH": {">=": "apple"}}` -- Comparators require int or float valued attribute, not string.
* `{"_": {"elmo_vector": {"IN": [0.0]}}}` -- `IN` requires either string or integer, extension attributes might return arbitrary types.
* I'm not really verifying that the predicates are valid (currently supported ones are `REGEX`, `IN, `NOT_IN`, comparators). If they pass something arbitrary there, the error might be confusing.
* I'm not checking that we don't get something like `{"ORTH": {"hello": "bye"}}`. I guess this wasn't really checked before.

I think we might consider separating the validation logic from the pattern parser? This might make it easier to write errors.

**Comment by ines** [comment]
> We should make it a subpackage. 

Yeah, then we could have `matcher/matcher.pyx`, `matcher/phrasematcher.pyx` and `matcher/entityruler.pyx`.

> We also need better pattern validation and errors.

Since we're already depending on `jsonschema` in `v2.x`, we could now validate this using a schema? If a pattern is added, it's validated by default (and the behaviour could be turned off by setting `validate=False` on intialization of the matcher, just in case.)

Here's a quick schema suggestion that should roughly describe the constraints (untested):

```json
{
  "$schema": "http://json-schema.org/draft-06/schema",
  "definitions": {
    "string_value": {
      "anyOf": [
        { 
          "type": "string" 
        },
        {
          "type": "object",
          "properties": {
            "REGEX": {
              "type": "string"
            },
            "IN": {
              "type": "array",
              "items": {
                "type": "string"
              }
            },
            "NOT_IN": {
              "type": "array",
              "items": {
                "type": "string"
              }
            }
          },
          "additionalProperties": false
        }
      ]
    },
    "integer_value": {
      "anyOf": [
        {
          "type": "integer",
        },
        {
          "type": "object",
          "properties": {
            "REGEX": {
              "type": "string"
            },
            "IN": {
              "type": "array",
              "items": {
                "type": "integer"
              }
            },
            "NOT_IN": {
              "type": "array",
              "items": {
                "type": "integer"
              }
            },
            "==": {
              "type": "integer"
            },
            ">=": {
              "type": "integer"
            },
            "<=": {
              "type": "integer"
            },
            ">": {
              "type": "integer"
            },
            "<": {
              "type": "integer"
            }
          },
          "additionalProperties": false
        }
      ]
    },
    "boolean_value": {
      "type": "boolean"
    },
    "underscore_value": {
      "anyOf": [
        {"type": ["string", "integer", "number", "array", "boolean", "null"]},
        {
          "type": "object",
          "properties": {
            "REGEX": {
              "type": "string"
            },
            "IN": {
              "type": ["string", "integer"]
            },
            "NOT_IN": {
              "type": ["string", "integer"]
            },
            "==": {
              "type": "integer"
            },
            ">=": {
              "type": "integer"
            },
            "<=": {
              "type": "integer"
            },
            ">": {
              "type": "integer"
            },
            "<": {
              "type": "integer"
            }
          },
          "additionalProperties": false
        }
      ]
    }
  },
  "type": "array",
  "items": {
    "type": "object",
    "properties": {
      "ORTH": {
        "title": "Verbatim token text",
        "$ref": "#/definitions/string_value"
      },
      "TEXT": {
        "title": "Verbatim token text (spaCy v2.1+)",
        "$ref": "#/definitions/string_value"
      },
      "LOWER": {
        "title": "Lowercase form of token text",
        "$ref": "#/definitions/string_value"
      },
      "POS": {
        "title": "Coarse-grained part-of-speech tag",
        "$ref": "#/definitions/string_value"
      },
      "TAG": {
        "title": "Fine-grained part-of-speech tag",
        "$ref": "#/definitions/string_value"
      },
      "DEP": {
        "title": "Dependency label",
        "$ref": "#/definitions/string_value"
      },
      "LEMMA": {
        "title": "Lemma (base form)",
        "$ref": "#/definitions/string_value"
      },
      "SHAPE": {
        "title": "Abstract token shape",
        "$ref": "#/definitions/string_value"
      },
      "ENT_TYPE": {
        "title": "Entity label of single token",
        "$ref": "#/definitions/string_value"
      },
      "LENGTH": {
        "title": "Token character length",
        "$ref": "#/definitions/integer_value"
      },
      "IS_ALPHA": {
        "title": "Token consists of alphanumeric characters",
        "$ref": "#/definitions/boolean_value"
      },
      "IS_ASCII": {
        "title": "Token consists of ASCII characters",
        "$ref": "#/definitions/boolean_value"
      },
      "IS_DIGIT": {
        "title": "Token consists of digits",
        "$ref": "#/definitions/boolean_value"
      },
      "IS_LOWER": {
        "title": "Token is lowercase",
        "$ref": "#/definitions/boolean_value"
      },
      "IS_UPPER": {
        "title": "Token is uppercase",
        "$ref": "#/definitions/boolean_value"
      },
      "IS_TITLE": {
        "title": "Token  is titlecase",
        "$ref": "#/definitions/boolean_value"
      },
      "IS_PUNCT": {
        "title": "Token is punctuation",
        "$ref": "#/definitions/boolean_value"
      },
      "IS_SPACE": {
        "title": "Token is whitespace",
        "$ref": "#/definitions/boolean_value"
      },
      "IS_STOP": {
        "title": "Token is stop word",
        "$ref": "#/definitions/boolean_value"
      },
      "LIKE_NUM": {
        "title": "Token resembles a number",
        "$ref": "#/definitions/boolean_value"
      },
      "LIKE_URL": {
        "title": "Token resembles a URL",
        "$ref": "#/definitions/boolean_value"
      },
      "LIKE_EMAIL": {
        "title": "Token resembles an email address",
        "$ref": "#/definitions/boolean_value"
      },
      "_": {
        "title": "Custom extension token attributes (token._.)",
        "type": "object",
        "properties": {
          "*": {
            "$ref": "#/definitions/underscore_value"
          }
        }
        
      }
    },
    "additionalProperties": false
  }
}
```

**Comment by honnibal** [comment]
@ines +1, looks good!

Do we want to merge this and reserve the reorg and schema validation for a subsequent PR? That'll let us get this out to the nightly.

Since we squash commits, I also want to avoid making the reorg in the same PR as these logic changes, since we won't be able to see the diff for this patch once we do the reorg.
